PackageFeedSortingTest functional test fails when < 10 packages in feed

The call to `responseText.IndexOf` does not account for the string not being found, due to reaching the end of the feed because the same package appears many times.

see: https://github.com/NuGet/NuGetGallery/blob/master/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs#L211

From https://nuget.visualstudio.com/NuGetBuild/_build?buildId=34967
```
2018-07-16T16:25:48.4421714Z       Expected string e2e.semver2prerel.180630.045128.5737195 to come before 63712.  Expected list is: webgrease, owin, xsockets.plugin.framework, basetestpackage, microbuild.core, test20180627163712, newtonsoft.json, e2e.semver2prerel.180629.000605.2522011, e2e.semver2prerel.180630.045128.5737195, 63712
2018-07-16T16:25:48.4421714Z       Expected: True
2018-07-16T16:25:48.4421714Z       Actual:   False
2018-07-16T16:25:48.4421714Z       Stack Trace:
2018-07-16T16:25:48.4421714Z         C:\vsts-agent\_work\51\s\tests\NuGetGallery.FunctionalTests\ODataFeeds\V2FeedExtendedTests.cs(233,0): at NuGetGallery.FunctionalTests.ODataFeeds.V2FeedExtendedTests.<PackageFeedSortingTest>d__11.MoveNext()
2018-07-16T16:25:48.4421714Z         --- End of stack trace from previous location where exception was thrown ---
2018-07-16T16:25:48.4421714Z            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-16T16:25:48.4421714Z            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-16T16:25:48.4421714Z         --- End of stack trace from previous location where exception was thrown ---
2018-07-16T16:25:48.4421714Z            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-16T16:25:48.4421714Z            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-16T16:25:48.4421714Z         --- End of stack trace from previous location where exception was thrown ---
2018-07-16T16:25:48.4421714Z            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-16T16:25:48.4421714Z            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-16T16:25:48.5047075Z ##[error]    NuGetGallery.FunctionalTests.ODataFeeds.V2FeedExtendedTests.PackageFeedSortingTest [FAIL]
```

--- a/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs
+++ b/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs
@@ -194,32 +194,25 @@ namespace NuGetGallery.FunctionalTests.ODataFeeds
         [Category("P1Tests")]
         public async Task PackageFeedSortingTest()
         {
-            var request = WebRequest.Create(UrlHelper.V2FeedRootUrl + @"stats/downloads/last6weeks/");
-            var response = await request.GetResponseAsync();
-
-            string responseText = await new StreamReader(response.GetResponseStream())
-                .ReadToEndAsync();
+            var topDownloadsResponse = await GetResponseTextAsync(UrlHelper.V2FeedRootUrl + @"stats/downloads/last6weeks/");
 
             // Search Gallery v2 feed for the top 10 unique package ids, or as many as exist.
-            string[] topPackages = GetTopPackageNames(responseText);
+            string[] topDownloadsNames = GetPackageNamesFromFeedResponse(topDownloadsResponse);
 
             // Ensure at least 1 package was found.
-            Assert.NotEmpty(topPackages);
+            Assert.NotEmpty(topDownloadsNames);
 
             // Search Gallery statistics for the top downloaded packages.
-            request = WebRequest.Create(UrlHelper.BaseUrl + @"stats/packageversions");
-            response = await request.GetResponseAsync();
-            responseText = await new StreamReader(response.GetResponseStream())
-                .ReadToEndAsync();
+            var statsResponse = await GetResponseTextAsync(UrlHelper.BaseUrl + @"stats/packageversions");
 
-            var expectedNames = String.Join(", ", topPackages);
-            var last = topPackages.First();
-            foreach (var current in topPackages.Skip(1))
+            var expectedNames = String.Join(", ", topDownloadsNames);
+            var last = topDownloadsNames.First();
+            foreach (var current in topDownloadsNames.Skip(1))
             {
                 // Check to make sure the top 10 packages are in the same order as the feed.
                 // We add angle brackets to prevent false failures due to duplicate package names in the page.
-                var condition = responseText.IndexOf(">" + last + "<", StringComparison.Ordinal)
-                    < responseText.IndexOf(">" + current + "<", StringComparison.Ordinal);
+                var condition = statsResponse.IndexOf(">" + last + "<", StringComparison.Ordinal)
+                    < statsResponse.IndexOf(">" + current + "<", StringComparison.Ordinal);
                 Assert.True(condition, $"Expected string {last} to come before {current}.  Expected list is: {expectedNames}.");
 
                 Assert.NotEqual(last, current);
@@ -227,12 +220,24 @@ namespace NuGetGallery.FunctionalTests.ODataFeeds
             }
         }
 
-        private string[] GetTopPackageNames(string feedResponseText)
+        private async Task<string> GetResponseTextAsync(string url)
         {
+            using (var wr = await WebRequest.Create(url).GetResponseAsync())
+            using (var sr = new StreamReader(wr.GetResponseStream()))
+            {
+                return await sr.ReadToEndAsync();
+            }
+        }
+
+        private string[] GetPackageNamesFromFeedResponse(string feedResponseText)
+        {
+            const string PackageIdStartKey = @"""PackageId"": """;
+            const string PackageIdEndKey = @"""";
+
             var results = new List<string>();
 
-            Func<string, int> seekStart = s => s.IndexOf(@"""PackageId"": """, StringComparison.Ordinal);
-            Func<string, int> seekEnd = s => s.IndexOf(@"""", StringComparison.Ordinal);
+            Func<string, int> seekStart = s => s.IndexOf(PackageIdStartKey, StringComparison.Ordinal);
+            Func<string, int> seekEnd = s => s.IndexOf(PackageIdEndKey, StringComparison.Ordinal);
 
             do
             {
@@ -241,9 +246,8 @@ namespace NuGetGallery.FunctionalTests.ODataFeeds
                 {
                     break;
                 }
-
-                // skip over `"PackageId": `
-                feedResponseText = feedResponseText.Substring(start + 14);
+                
+                feedResponseText = feedResponseText.Substring(start + PackageIdStartKey.Length);
                 var end = seekEnd(feedResponseText);
                 if (end >= 0)
                 {

--- a/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs
+++ b/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs
@@ -221,6 +221,9 @@ namespace NuGetGallery.FunctionalTests.ODataFeeds
                 var condition = responseText.IndexOf(">" + last + "<", StringComparison.Ordinal)
                     < responseText.IndexOf(">" + current + "<", StringComparison.Ordinal);
                 Assert.True(condition, $"Expected string {last} to come before {current}.  Expected list is: {expectedNames}.");
+
+                Assert.NotEqual(last, current);
+                last = current;
             }
         }
 

--- a/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs
+++ b/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs
@@ -7,6 +7,7 @@ using System.ComponentModel;
 using System.IO;
 using System.Linq;
 using System.Net;
+using System.Text;
 using System.Threading.Tasks;
 using Newtonsoft.Json.Linq;
 using NuGetGallery.FunctionalTests.Helpers;
@@ -196,47 +197,64 @@ namespace NuGetGallery.FunctionalTests.ODataFeeds
             var request = WebRequest.Create(UrlHelper.V2FeedRootUrl + @"stats/downloads/last6weeks/");
             var response = await request.GetResponseAsync();
 
-            string responseText;
-            using (var sr = new StreamReader(response.GetResponseStream()))
-            {
-                responseText = await sr.ReadToEndAsync();
-            }
+            string responseText = await new StreamReader(response.GetResponseStream())
+                .ReadToEndAsync();
 
-            // Grab the top 10 package names in the feed.
-            string[] packageName = new string[10];
-            responseText = packageName[0] = responseText.Substring(responseText.IndexOf(@"""PackageId"": """, StringComparison.Ordinal) + 14);
-            packageName[0] = packageName[0].Substring(0, responseText.IndexOf(@"""", StringComparison.Ordinal));
-            for (int i = 1; i < 10; i++)
-            {
-                responseText = packageName[i] = responseText.Substring(responseText.IndexOf(@"""PackageId"": """, StringComparison.Ordinal) + 14);
-                packageName[i] = packageName[i].Substring(0, responseText.IndexOf(@"""", StringComparison.Ordinal));
-                // Sometimes two versions of a single package appear in the top 10.  Stripping second and later instances for this test.
-                for (int j = 0; j < i; j++)
-                {
-                    if (packageName[j] == packageName[i])
-                    {
-                        packageName[i] = null;
-                        i--;
-                    }
-                }
-            }
+            // Search Gallery v2 feed for the top 10 unique package ids, or as many as exist.
+            string[] topPackages = GetTopPackageNames(responseText);
 
+            // Ensure at least 1 package was found.
+            Assert.NotEmpty(topPackages);
+
+            // Search Gallery statistics for the top downloaded packages.
             request = WebRequest.Create(UrlHelper.BaseUrl + @"stats/packageversions");
-
-            // Get the response.
             response = await request.GetResponseAsync();
-            using (var sr = new StreamReader(response.GetResponseStream()))
-            {
-                responseText = await sr.ReadToEndAsync();
-            }
+            responseText = await new StreamReader(response.GetResponseStream())
+                .ReadToEndAsync();
 
-            for (int i = 1; i < 10; i++)
+            var expectedNames = String.Join(", ", topPackages);
+            var last = topPackages.First();
+            foreach (var current in topPackages.Skip(1))
             {
                 // Check to make sure the top 10 packages are in the same order as the feed.
                 // We add angle brackets to prevent false failures due to duplicate package names in the page.
-                var condition = responseText.IndexOf(">" + packageName[i - 1] + "<", StringComparison.Ordinal) < responseText.IndexOf(">" + packageName[i] + "<", StringComparison.Ordinal);
-                Assert.True(condition, "Expected string " + packageName[i - 1] + " to come before " + packageName[i] + ".  Expected list is: " + packageName[0] + ", " + packageName[1] + ", " + packageName[2] + ", " + packageName[3] + ", " + packageName[4] + ", " + packageName[5] + ", " + packageName[6] + ", " + packageName[7] + ", " + packageName[8] + ", " + packageName[9]);
+                var condition = responseText.IndexOf(">" + last + "<", StringComparison.Ordinal)
+                    < responseText.IndexOf(">" + current + "<", StringComparison.Ordinal);
+                Assert.True(condition, $"Expected string {last} to come before {current}.  Expected list is: {expectedNames}.");
             }
+        }
+
+        private string[] GetTopPackageNames(string feedResponseText)
+        {
+            var results = new List<string>();
+
+            Func<string, int> seekStart = s => s.IndexOf(@"""PackageId"": """, StringComparison.Ordinal);
+            Func<string, int> seekEnd = s => s.IndexOf(@"""", StringComparison.Ordinal);
+
+            do
+            {
+                var start = seekStart(feedResponseText);
+                if (start < 0)
+                {
+                    break;
+                }
+
+                // skip over `"PackageId": `
+                feedResponseText = feedResponseText.Substring(start + 14);
+                var end = seekEnd(feedResponseText);
+                if (end >= 0)
+                {
+                    var name = feedResponseText.Substring(0, end);
+                    if (!results.Contains(name, StringComparer.Ordinal))
+                    {
+                        results.Add(name);
+                    }
+                    feedResponseText = feedResponseText.Substring(end);
+                }
+            }
+            while (results.Count < 10);
+
+            return results.ToArray();
         }
     }
 }

